File: 1046-max-consecutive-ones-iii/1046-max-consecutive-ones-iii.go
func longestOnes(nums []int, k int) int {
    maxlen, zero, l, r := 0, 0, 0, 0

    for r < len(nums) {
        if nums[r] == 0 {
            zero += 1
        }

        if zero > k {
            if nums[l] == 0 {
                zero -= 1
            }
            l += 1
        }

        if zero <= k {
            maxlen = max(maxlen, r-l+1)
        }

        r += 1
    }

    return maxlen
}

File: 1046-max-consecutive-ones-iii/1046-max-consecutive-ones-iii.ts
function longestOnes(nums: number[], k: number): number {
    let maxLen = 0, zero = 0, l = 0;

    for (let r = 0; r < nums.length; r++) {
        if (nums[r] === 0) zero++;
        while (zero > k) if (nums[l++] === 0) zero--;
        maxLen = Math.max(maxLen, r - l + 1);
    }
    
    return maxLen;
}
